File: Problem139_task-scheduler.java
// Time Complexity: O(n)
// Space Complexity: O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    public int leastInterval(char[] tasks, int n) {
        if(tasks == null || tasks.length == 0) {
            return 0;
        }
        HashMap<Character, Integer> map = new HashMap<>();
        int maxFreq = 0;
        for(int i = 0; i < tasks.length; i++) {
            map.put(tasks[i], map.getOrDefault(tasks[i], 0) + 1);
            maxFreq = Math.max(maxFreq, map.get(tasks[i]));
        }
        int maxCount = 0;
        for(char task: map.keySet()) {
            if(maxFreq == map.get(task)) {
                maxCount++;
            }
        }
        int partitions = maxFreq - 1;
        int empty = partitions * (n - (maxCount - 1));
        int pendingTasks = tasks.length - (maxCount * maxFreq);
        int idle = Math.max(0, empty - pendingTasks);
        
        return tasks.length + idle;
    }
}

File: Problem-139 Task Scheduler.py
# 621. Task Scheduler
# https://leetcode.com/problems/task-scheduler/

# Time Complexity: O(n)
# Space Complexity: O(1) [only 26 letters possible so O(26)]

class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        freq_hashmap = dict()
        max_count = 0
        
        for i in tasks:
            if i in freq_hashmap:
                freq_hashmap[i] += 1
            else:
                freq_hashmap[i] = 1
            
            max_count = max(max_count, freq_hashmap[i])